# **Assignment 1 - Part C**


## Assignment 1 - Regression Models
Part C: Experiment on multivariate linear regression with feature engineering or with any algorithms of your choice.

Training and testing dataset has been cleaned in the part A and stored in this google drive link: https://drive.google.com/drive/folders/1_50s_IxJsABTRFTWBR0XjuSOnDL4YFuQ

The steps are:
1.  Load and Explore Dataset
2.  Data Preparation and Feature Engineering
    * Handling missing values (completed in part A of the assignment) 
    * Ordinal Encoding
    * Feature creation
    * Handling outliners
    * Removing redundant or irrelevant features
3.  Assess Baseline Model
4.  Train Initial Model with all Features
5.  Train Lasso Model with the all features
6.  Train Ridge Model with the all features
7.  Train Elasticnet Model with the all features


### 1. Load and Explore Dataset

**[1.1]** Import the pandas and numpy packages

In [101]:
import pandas as pd
import numpy as np
import altair as alt

**[1.2]** Mounting the drive

In this notebook, I'm mounting the Google drive to read a csv file that is stored on my drive. You must allow access to your drive by signing in to your Google account.

In [102]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [103]:
# After executing the cell above, Drive files will be present in "/content/drive/My Drive". The below command lists the contents in the drive:
!ls "/content/drive/My Drive"

 AT2_default_template.gdoc
 CNN_Articles_2021-2023.csv
 CNN_Articles_2021-2023.gsheet
'Colab Notebooks'
'Copy of Data meaning https:  tinyurl.com DSI-Dmean-c (1).gsheet'
'Copy of Data meaning https:  tinyurl.com DSI-Dmean-c (2).gsheet'
'Copy of Data meaning https:  tinyurl.com DSI-Dmean-c (3).gsheet'
'Copy of Data meaning https:  tinyurl.com DSI-Dmean-c.gsheet'
'Copy of MLAA - Lecture 1.gslides'
'Copy of MLAA - Lecture 2.gslides'
'Copy of MLAA - Lecture 5.gslides'
 data
'DSP_AT3 _REPORT.gdoc'
'DSP - Lab 4 - Exercise 1 - Solutions.gdoc'
 DVN_data
'Experiment reports'
 Lambda_Project_Report_Draft.gdoc
 NLP_data
'ODI Data Ethics Canvas.gdoc'
 Untitled
'Untitled document (1).gdoc'
'Untitled document.gdoc'


**[1.3]** Create a variable called file_url containing the link to the CSV file for training data

In [104]:
file_url = pd.read_csv('/content/drive/My Drive/data/training_cleaned2.csv')

**[1.4]** Load the training dataset into dataframe called df

In [105]:
df = pd.DataFrame(file_url)

**[1.5]** Display the first 5 rows of df

In [106]:
df.head()

Unnamed: 0  avgAnnCount  avgDeathsPerYear  TARGET_deathRate  incidenceRate  \
0           0    88.000000                40             261.0     561.400000   
1           1    73.000000                35             167.3     345.600000   
2           2   292.000000               124             191.0     468.400000   
3           3  1962.667684                 7             165.4     453.549422   
4           4    43.000000                20             160.6     349.700000   

   medIncome  popEst2015  povertyPercent  studyPerCap           binnedInc  \
0      29090       13352            26.8  2771.120431    [22640, 34218.1]   
1      29782       21903            38.8     0.000000    [22640, 34218.1]   
2      41955       48985            15.5     0.000000  (40362.7, 42724.4]   
3      55378        3007            11.1     0.000000  (54545.6, 61494.5]   
4      26309        8551            35.3     0.000000    [22640, 34218.1]   

   ...  PctEmpPrivCoverage  PctPublicCoverage  PctPublicCoverageAlone  \
0  ...                32.0               47.5                    32.9   
1  ...                18.8               45.3                    34.1   
2  ...                44.9               34.5                    16.0   
3  ...                49.6               30.1                    15.2   
4  ...                30.4               45.1                    24.5   

    PctWhite   PctBlack  PctAsian  PctOtherRace  PctMarriedHouseholds  \
0  99.693045   0.044920  0.000000      0.000000             55.499459   
1  94.791383   1.649850  0.063631      2.854286             52.818296   
2  95.102348   1.741749  0.376429      0.445611             50.560800   
3  85.833870   0.933677  0.160979      7.244044             52.565181   
4  24.535525  73.223736  0.394100      1.396239             33.641208   

   BirthRate  Id  
0   6.838710   0  
1   4.799131   1  
2   3.996826   2  
3   3.291536   3  
4   3.166561   4  

[5 rows x 34 columns]

**[1.6]** Display the last 5 rows of df

In [107]:
df.tail()

Unnamed: 0  avgAnnCount  avgDeathsPerYear  TARGET_deathRate  \
2433        2433   389.000000               157             185.3   
2434        2434   286.000000               117             196.7   
2435        2435   103.000000                42             204.1   
2436        2436  1962.667684                23             171.1   
2437        2437   219.000000                89             179.7   

      incidenceRate  medIncome  popEst2015  povertyPercent  studyPerCap  \
2433     456.600000      65485       83199             7.7     0.000000   
2434     492.400000      42477       46222            16.9   281.251352   
2435     506.700000      40339       18201            21.3     0.000000   
2436     453.549422      39764        8856            16.7     0.000000   
2437     449.100000      40801       37585            17.2     0.000000   

               binnedInc  ...  PctEmpPrivCoverage  PctPublicCoverage  \
2433   (61494.5, 125635]  ...                55.1               24.2   
2434  (40362.7, 42724.4]  ...                46.2               34.9   
2435  (37413.8, 40362.7]  ...                34.4               36.9   
2436  (37413.8, 40362.7]  ...                37.3               40.0   
2437  (40362.7, 42724.4]  ...                37.5               39.4   

      PctPublicCoverageAlone   PctWhite   PctBlack  PctAsian  PctOtherRace  \
2433                    11.3  91.187608   4.445537  1.405213      0.858126   
2434                    18.8  90.130702   5.943936  0.472935      0.485833   
2435                    20.6  65.463178  30.550955  0.645909      0.104891   
2436                    21.3  94.625317   0.154508  0.629070      0.684251   
2437                    21.8  91.393432   3.381528  0.055305      3.149772   

      PctMarriedHouseholds  BirthRate    Id  
2433             60.095060   5.175689  3041  
2434             51.648588   4.651829  3042  
2435             49.758980   4.344104  3043  
2436             49.880605   6.210826  3045  
2437             49.892641   6.549178  3046  

[5 rows x 34 columns]

**[1.7]** Display the dimensions (shape) of df

In [108]:
df.shape

(2438, 34)

**[1.8]** Display the list of columns of df

In [109]:
df.columns


Index(['Unnamed: 0', 'avgAnnCount', 'avgDeathsPerYear', 'TARGET_deathRate',
       'incidenceRate', 'medIncome', 'popEst2015', 'povertyPercent',
       'studyPerCap', 'binnedInc', 'MedianAge', 'MedianAgeMale',
       'MedianAgeFemale', 'Geography', 'AvgHouseholdSize', 'PercentMarried',
       'PctNoHS18_24', 'PctHS18_24', 'PctBachDeg18_24', 'PctHS25_Over',
       'PctBachDeg25_Over', 'PctEmployed16_Over', 'PctUnemployed16_Over',
       'PctPrivateCoverage', 'PctEmpPrivCoverage', 'PctPublicCoverage',
       'PctPublicCoverageAlone', 'PctWhite', 'PctBlack', 'PctAsian',
       'PctOtherRace', 'PctMarriedHouseholds', 'BirthRate', 'Id'],
      dtype='object')

**[1.9]** Display the descriptive statistics of df


In [110]:
df.describe()

Unnamed: 0   avgAnnCount  avgDeathsPerYear  TARGET_deathRate  \
count  2438.000000   2438.000000       2438.000000       2438.000000   
mean   1218.500000    587.171930        180.592699        178.848482   
std     703.934301   1236.447895        438.473144         27.538452   
min       0.000000      6.000000          3.000000         66.300000   
25%     609.250000     76.000000         28.000000        161.400000   
50%    1218.500000    172.500000         62.000000        178.100000   
75%    1827.750000    521.500000        151.000000        195.300000   
max    2437.000000  24965.000000       9445.000000        293.900000   

       incidenceRate      medIncome    popEst2015  povertyPercent  \
count    2438.000000    2438.000000  2.438000e+03     2438.000000   
mean      448.308247   47028.427810  9.851579e+04       16.854348   
std        53.247868   11919.386622  2.745272e+05        6.388398   
min       201.300000   22640.000000  8.270000e+02        3.200000   
25%       420.300000   38872.750000  1.154575e+04       12.100000   
50%       453.549422   45186.500000  2.694250e+04       15.900000   
75%       481.975000   52492.500000  6.952450e+04       20.400000   
max      1014.200000  125635.000000  5.238216e+06       47.400000   

       studyPerCap    MedianAge  ...  PctEmpPrivCoverage  PctPublicCoverage  \
count  2438.000000  2438.000000  ...         2438.000000        2438.000000   
mean    166.019192    40.937707  ...           41.242453          36.275718   
std     563.813830     5.190141  ...            9.366433           7.823347   
min       0.000000    22.300000  ...           13.500000          11.200000   
25%       0.000000    37.800000  ...           34.600000          30.925000   
50%       0.000000    41.000000  ...           41.300000          36.400000   
75%      92.562203    44.075000  ...           47.600000          41.500000   
max    9762.308998    65.300000  ...           70.700000          65.100000   

       PctPublicCoverageAlone     PctWhite     PctBlack     PctAsian  \
count             2438.000000  2438.000000  2438.000000  2438.000000   
mean                19.234824    83.776008     9.043586     1.230949   
std                  6.094083    16.275962    14.361902     2.558111   
min                  2.600000    10.199155     0.000000     0.000000   
25%                 15.000000    77.341176     0.634412     0.247022   
50%                 18.800000    90.121405     2.302054     0.550713   
75%                 23.100000    95.464864    10.452697     1.207158   
max                 46.600000   100.000000    84.866024    42.619425   

       PctOtherRace  PctMarriedHouseholds    BirthRate           Id  
count   2438.000000           2438.000000  2438.000000  2438.000000  
mean       1.947961             51.216669     5.627447  1523.789171  
std        3.541698              6.501254     1.966257   874.905673  
min        0.000000             22.992490     0.000000     0.000000  
25%        0.288937             47.826087     4.516355   775.250000  
50%        0.798971             51.662077     5.366485  1512.500000  
75%        2.106155             55.331622     6.458322  2279.750000  
max       41.930251             78.075397    18.556701  3046.000000  

[8 rows x 32 columns]

**[1.10]** Display the summary (info) of df

In [111]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2438 entries, 0 to 2437
Data columns (total 34 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Unnamed: 0              2438 non-null   int64  
 1   avgAnnCount             2438 non-null   float64
 2   avgDeathsPerYear        2438 non-null   int64  
 3   TARGET_deathRate        2438 non-null   float64
 4   incidenceRate           2438 non-null   float64
 5   medIncome               2438 non-null   int64  
 6   popEst2015              2438 non-null   int64  
 7   povertyPercent          2438 non-null   float64
 8   studyPerCap             2438 non-null   float64
 9   binnedInc               2438 non-null   object 
 10  MedianAge               2438 non-null   float64
 11  MedianAgeMale           2438 non-null   float64
 12  MedianAgeFemale         2438 non-null   float64
 13  Geography               2438 non-null   object 
 14  AvgHouseholdSize        2438 non-null   

**[1.11]** Create a variable called file_url_test containing the link to the CSV file for testing data

In [112]:
file_url_test = pd.read_csv('/content/drive/My Drive/data/testing_cleaned2.csv')

**[1.12]** Load the testing dataset into dataframe called df_test

In [113]:
df_test = pd.DataFrame(file_url_test)

**[1.13]** Display the first 5 rows of df_test

In [114]:
df_test.head()

Unnamed: 0  avgAnnCount  avgDeathsPerYear  TARGET_deathRate  incidenceRate  \
0           0        449.0               154             159.5          479.8   
1           1        340.0               140             167.2          438.5   
2           2         54.0                18             131.6          410.8   
3           3         94.0                46             189.4          403.8   
4           4       2718.0              1065             168.9          432.1   

   medIncome  popEst2015  povertyPercent  studyPerCap           binnedInc  \
0      51880      104926            18.7    57.183158  (51046.4, 54545.6]   
1      55472       55423            12.4     0.000000  (54545.6, 61494.5]   
2      49380       10103            11.7     0.000000  (48021.6, 51046.4]   
3      45979       16708            13.5   598.515681    (45201, 48021.6]   
4      51527      726106            20.7    60.597213  (51046.4, 54545.6]   

   ...  PctEmpPrivCoverage  PctPublicCoverage  PctPublicCoverageAlone  \
0  ...                51.0               24.9                    13.1   
1  ...                37.6               36.3                    16.0   
2  ...                32.6               40.3                    19.4   
3  ...                45.1               33.1                    13.3   
4  ...                41.7               37.7                    25.8   

    PctWhite  PctBlack   PctAsian  PctOtherRace  PctMarriedHouseholds  \
0  81.260411  4.154831  10.045737      0.876222             41.071243   
1  93.660078  0.818115   0.626281      3.116360             57.529142   
2  98.292181  0.041152   0.164609      0.051440             55.928482   
3  96.090377  1.555569   0.715680      0.378541             48.409405   
4  57.002148  7.093743  14.785464     11.692122             51.852122   

   BirthRate    Id  
0   4.367123  2553  
1   6.844366   904  
2   1.604585  2192  
3   8.255410  1326  
4   6.148433  2394  

[5 rows x 34 columns]

**[1.14]** Display the last 5 rows of df_test

In [115]:
df_test.tail()

Unnamed: 0  avgAnnCount  avgDeathsPerYear  TARGET_deathRate  \
604         604    27.000000                14             187.6   
605         605    30.000000                 9             131.6   
606         606   583.000000               258             187.5   
607         607  1962.667684                31             174.2   
608         608    42.000000                17             199.1   

     incidenceRate  medIncome  popEst2015  povertyPercent  studyPerCap  \
604     415.700000      50155        4769            12.0     0.000000   
605     444.600000      46961        4854            14.0     0.000000   
606     429.400000      39907      127780            22.1    62.607607   
607     453.549422      50905       14219             9.3     0.000000   
608     491.800000      26408        7507            38.5     0.000000   

              binnedInc  ...  PctEmpPrivCoverage  PctPublicCoverage  \
604  (48021.6, 51046.4]  ...                42.0               27.4   
605    (45201, 48021.6]  ...                40.7               30.9   
606  (37413.8, 40362.7]  ...                39.8               37.9   
607  (48021.6, 51046.4]  ...                52.1               33.9   
608    [22640, 34218.1]  ...                28.0               52.2   

     PctPublicCoverageAlone   PctWhite   PctBlack  PctAsian  PctOtherRace  \
604                    12.1  85.768985   0.616890  0.042544      2.105935   
605                    15.8  96.122281   0.800164  0.307755      1.292573   
606                    22.7  81.407683   6.285701  4.388991      2.247924   
607                    17.6  93.756201   1.658398  0.836286      0.290574   
608                    33.4  13.933562  85.947799  0.118640      0.000000   

     PctMarriedHouseholds  BirthRate    Id  
604             51.104816   6.659013  2247  
605             49.671883   5.760870  2947  
606             47.875108   6.387886  1746  
607             45.219595   5.893846  1822  
608             25.540275   4.610131    92  

[5 rows x 34 columns]

**[1.15]** Display the dimensions (shape) of df_test


In [116]:
df_test.shape

(609, 34)

**[1.16]** Display the list of columns of df_test

In [117]:
df_test.columns

Index(['Unnamed: 0', 'avgAnnCount', 'avgDeathsPerYear', 'TARGET_deathRate',
       'incidenceRate', 'medIncome', 'popEst2015', 'povertyPercent',
       'studyPerCap', 'binnedInc', 'MedianAge', 'MedianAgeMale',
       'MedianAgeFemale', 'Geography', 'AvgHouseholdSize', 'PercentMarried',
       'PctNoHS18_24', 'PctHS18_24', 'PctBachDeg18_24', 'PctHS25_Over',
       'PctBachDeg25_Over', 'PctEmployed16_Over', 'PctUnemployed16_Over',
       'PctPrivateCoverage', 'PctEmpPrivCoverage', 'PctPublicCoverage',
       'PctPublicCoverageAlone', 'PctWhite', 'PctBlack', 'PctAsian',
       'PctOtherRace', 'PctMarriedHouseholds', 'BirthRate', 'Id'],
      dtype='object')

**[1.17]** Display the descriptive statistics of df_test

---



In [118]:
df_test.describe()

Unnamed: 0   avgAnnCount  avgDeathsPerYear  TARGET_deathRate  \
count  609.000000    609.000000        609.000000        609.000000   
mean   304.000000    683.067944        207.476190        177.925780   
std    175.947435   1978.627131        708.568892         28.599788   
min      0.000000      9.000000          3.000000         59.700000   
25%    152.000000     75.000000         28.000000        160.200000   
50%    304.000000    169.000000         61.000000        178.200000   
75%    456.000000    513.000000        138.000000        194.100000   
max    608.000000  38150.000000      14010.000000        362.800000   

       incidenceRate      medIncome    popEst2015  povertyPercent  \
count     609.000000     609.000000  6.090000e+02      609.000000   
mean      448.109810   47202.812808  1.191372e+05       16.973563   
std        59.573942   12520.769346  4.899460e+05        6.495675   
min       269.900000   24035.000000  1.358000e+03        4.700000   
25%       420.500000   39025.000000  1.218100e+04       12.400000   
50%       453.549422   45292.000000  2.558600e+04       15.500000   
75%       476.700000   52476.000000  6.783700e+04       20.400000   
max      1206.900000  106871.000000  1.017029e+07       41.900000   

       studyPerCap   MedianAge  ...  PctEmpPrivCoverage  PctPublicCoverage  \
count   609.000000  609.000000  ...          609.000000         609.000000   
mean    112.885432   40.587255  ...           41.011658          36.160263   
std     358.996110    5.183762  ...            9.771919           7.920750   
min       0.000000   23.200000  ...           16.300000          14.800000   
25%       0.000000   37.400000  ...           33.600000          30.600000   
50%       0.000000   40.800000  ...           40.700000          36.300000   
75%      58.267135   43.900000  ...           47.800000          41.800000   
max    4260.367794   56.500000  ...           70.200000          57.500000   

       PctPublicCoverageAlone    PctWhite    PctBlack    PctAsian  \
count              609.000000  609.000000  609.000000  609.000000   
mean                19.261084   83.121969    9.365754    1.346103   
std                  6.193365   16.793593   15.215528    2.809767   
min                  6.200000   13.933562    0.000000    0.000000   
25%                 14.300000   77.048638    0.607010    0.278340   
50%                 18.900000   89.606996    2.004479    0.547264   
75%                 23.100000   95.363876   11.063967    1.245240   
max                 38.800000   99.498433   85.947799   33.760905   

       PctOtherRace  PctMarriedHouseholds   BirthRate           Id  
count    609.000000            609.000000  609.000000   609.000000  
mean       2.125887             51.352773    5.691785  1519.840722  
std        3.419138              6.856244    2.063121   899.544555  
min        0.000000             23.885628    0.284091    10.000000  
25%        0.318345             47.399630    4.585307   704.000000  
50%        0.943696             51.708723    5.475268  1597.000000  
75%        2.356943             55.590179    6.593969  2298.000000  
max       25.504910             71.703057   21.326165  3044.000000  

[8 rows x 32 columns]

**[1.18]** Display the summary (info) of df_test

In [119]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 609 entries, 0 to 608
Data columns (total 34 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Unnamed: 0              609 non-null    int64  
 1   avgAnnCount             609 non-null    float64
 2   avgDeathsPerYear        609 non-null    int64  
 3   TARGET_deathRate        609 non-null    float64
 4   incidenceRate           609 non-null    float64
 5   medIncome               609 non-null    int64  
 6   popEst2015              609 non-null    int64  
 7   povertyPercent          609 non-null    float64
 8   studyPerCap             609 non-null    float64
 9   binnedInc               609 non-null    object 
 10  MedianAge               609 non-null    float64
 11  MedianAgeMale           609 non-null    float64
 12  MedianAgeFemale         609 non-null    float64
 13  Geography               609 non-null    object 
 14  AvgHouseholdSize        609 non-null    fl

### 2. Data Preparation & Feature Engineering

**[2.1]** Let's create a copy of our training dataframe

In [120]:
df_cleaned = df.copy()

**[2.2]** Let's create a copy of our testing dataframe

In [121]:
df_test_cleaned = df_test.copy()

**[2.3]** Ordinal encoding: This method is useful for ordinal categorical features, such as "binnedInc", where the categories have an inherent order. It involves assigning a numerical value to each category based on its order. For example, we could assign the value 1 to the lowest income bracket and 5 to the highest income bracket

For training set

In [122]:
print(df_cleaned['binnedInc'].unique())

['[22640, 34218.1]' '(40362.7, 42724.4]' '(54545.6, 61494.5]'
 '(61494.5, 125635]' '(34218.1, 37413.8]' '(48021.6, 51046.4]'
 '(42724.4, 45201]' '(45201, 48021.6]' '(51046.4, 54545.6]'
 '(37413.8, 40362.7]']


In [123]:
from sklearn.preprocessing import OrdinalEncoder

# Create an ordinal encoder instance
encoder = OrdinalEncoder(categories=[['[22640, 34218.1]', '(34218.1, 37413.8]', 
                                      '(37413.8, 40362.7]', '(40362.7, 42724.4]', 
                                      '(42724.4, 45201]', '(45201, 48021.6]', 
                                      '(48021.6, 51046.4]', '(51046.4, 54545.6]', 
                                      '(54545.6, 61494.5]', '(61494.5, 125635]', 
                                      'Unknown']])

# Replace unknown categories with 'Unknown'
df_cleaned = df.replace(['?'], 'Unknown')

# Apply the encoder to the 'binnedInc' column
df_cleaned ['binnedInc_ordinal'] = encoder.fit_transform(df_cleaned[['binnedInc']])



In [124]:
df_cleaned

Unnamed: 0  avgAnnCount  avgDeathsPerYear  TARGET_deathRate  \
0              0    88.000000                40             261.0   
1              1    73.000000                35             167.3   
2              2   292.000000               124             191.0   
3              3  1962.667684                 7             165.4   
4              4    43.000000                20             160.6   
...          ...          ...               ...               ...   
2433        2433   389.000000               157             185.3   
2434        2434   286.000000               117             196.7   
2435        2435   103.000000                42             204.1   
2436        2436  1962.667684                23             171.1   
2437        2437   219.000000                89             179.7   

      incidenceRate  medIncome  popEst2015  povertyPercent  studyPerCap  \
0        561.400000      29090       13352            26.8  2771.120431   
1        345.600000      29782       21903            38.8     0.000000   
2        468.400000      41955       48985            15.5     0.000000   
3        453.549422      55378        3007            11.1     0.000000   
4        349.700000      26309        8551            35.3     0.000000   
...             ...        ...         ...             ...          ...   
2433     456.600000      65485       83199             7.7     0.000000   
2434     492.400000      42477       46222            16.9   281.251352   
2435     506.700000      40339       18201            21.3     0.000000   
2436     453.549422      39764        8856            16.7     0.000000   
2437     449.100000      40801       37585            17.2     0.000000   

               binnedInc  ...  PctPublicCoverage  PctPublicCoverageAlone  \
0       [22640, 34218.1]  ...               47.5                    32.9   
1       [22640, 34218.1]  ...               45.3                    34.1   
2     (40362.7, 42724.4]  ...               34.5                    16.0   
3     (54545.6, 61494.5]  ...               30.1                    15.2   
4       [22640, 34218.1]  ...               45.1                    24.5   
...                  ...  ...                ...                     ...   
2433   (61494.5, 125635]  ...               24.2                    11.3   
2434  (40362.7, 42724.4]  ...               34.9                    18.8   
2435  (37413.8, 40362.7]  ...               36.9                    20.6   
2436  (37413.8, 40362.7]  ...               40.0                    21.3   
2437  (40362.7, 42724.4]  ...               39.4                    21.8   

       PctWhite   PctBlack  PctAsian  PctOtherRace  PctMarriedHouseholds  \
0     99.693045   0.044920  0.000000      0.000000             55.499459   
1     94.791383   1.649850  0.063631      2.854286             52.818296   
2     95.102348   1.741749  0.376429      0.445611             50.560800   
3     85.833870   0.933677  0.160979      7.244044             52.565181   
4     24.535525  73.223736  0.394100      1.396239             33.641208   
...         ...        ...       ...           ...                   ...   
2433  91.187608   4.445537  1.405213      0.858126             60.095060   
2434  90.130702   5.943936  0.472935      0.485833             51.648588   
2435  65.463178  30.550955  0.645909      0.104891             49.758980   
2436  94.625317   0.154508  0.629070      0.684251             49.880605   
2437  91.393432   3.381528  0.055305      3.149772             49.892641   

      BirthRate    Id  binnedInc_ordinal  
0      6.838710     0                0.0  
1      4.799131     1                0.0  
2      3.996826     2                3.0  
3      3.291536     3                8.0  
4      3.166561     4                0.0  
...         ...   ...                ...  
2433   5.175689  3041                9.0  
2434   4.651829  3042                3.0  
2435   4.344104  3043                2.0  
2436   6.210826  304

For testing set

In [125]:
print(df_test_cleaned['binnedInc'].unique())

['(51046.4, 54545.6]' '(54545.6, 61494.5]' '(48021.6, 51046.4]'
 '(45201, 48021.6]' '(42724.4, 45201]' '(40362.7, 42724.4]'
 '(61494.5, 125635]' '[22640, 34218.1]' '(37413.8, 40362.7]'
 '(34218.1, 37413.8]']


In [126]:
from sklearn.preprocessing import OrdinalEncoder

# Create an ordinal encoder instance
encoder = OrdinalEncoder(categories=[['[22640, 34218.1]', '(34218.1, 37413.8]', 
                                      '(37413.8, 40362.7]', '(40362.7, 42724.4]',
                                      '(42724.4, 45201]', '(45201, 48021.6]', 
                                      '(48021.6, 51046.4]', '(51046.4, 54545.6]',
                                      '(54545.6, 61494.5]', '(61494.5, 125635]']])

# Replace unknown categories with 'Unknown'
df_test_cleaned = df.replace(['?'], 'Unknown')

# Apply the encoder to the 'binnedInc' column
df_test_cleaned['binnedInc_ordinal'] = encoder.fit_transform(df_test_cleaned[['binnedInc']])


In [127]:
df_test_cleaned

Unnamed: 0  avgAnnCount  avgDeathsPerYear  TARGET_deathRate  \
0              0    88.000000                40             261.0   
1              1    73.000000                35             167.3   
2              2   292.000000               124             191.0   
3              3  1962.667684                 7             165.4   
4              4    43.000000                20             160.6   
...          ...          ...               ...               ...   
2433        2433   389.000000               157             185.3   
2434        2434   286.000000               117             196.7   
2435        2435   103.000000                42             204.1   
2436        2436  1962.667684                23             171.1   
2437        2437   219.000000                89             179.7   

      incidenceRate  medIncome  popEst2015  povertyPercent  studyPerCap  \
0        561.400000      29090       13352            26.8  2771.120431   
1        345.600000      29782       21903            38.8     0.000000   
2        468.400000      41955       48985            15.5     0.000000   
3        453.549422      55378        3007            11.1     0.000000   
4        349.700000      26309        8551            35.3     0.000000   
...             ...        ...         ...             ...          ...   
2433     456.600000      65485       83199             7.7     0.000000   
2434     492.400000      42477       46222            16.9   281.251352   
2435     506.700000      40339       18201            21.3     0.000000   
2436     453.549422      39764        8856            16.7     0.000000   
2437     449.100000      40801       37585            17.2     0.000000   

               binnedInc  ...  PctPublicCoverage  PctPublicCoverageAlone  \
0       [22640, 34218.1]  ...               47.5                    32.9   
1       [22640, 34218.1]  ...               45.3                    34.1   
2     (40362.7, 42724.4]  ...               34.5                    16.0   
3     (54545.6, 61494.5]  ...               30.1                    15.2   
4       [22640, 34218.1]  ...               45.1                    24.5   
...                  ...  ...                ...                     ...   
2433   (61494.5, 125635]  ...               24.2                    11.3   
2434  (40362.7, 42724.4]  ...               34.9                    18.8   
2435  (37413.8, 40362.7]  ...               36.9                    20.6   
2436  (37413.8, 40362.7]  ...               40.0                    21.3   
2437  (40362.7, 42724.4]  ...               39.4                    21.8   

       PctWhite   PctBlack  PctAsian  PctOtherRace  PctMarriedHouseholds  \
0     99.693045   0.044920  0.000000      0.000000             55.499459   
1     94.791383   1.649850  0.063631      2.854286             52.818296   
2     95.102348   1.741749  0.376429      0.445611             50.560800   
3     85.833870   0.933677  0.160979      7.244044             52.565181   
4     24.535525  73.223736  0.394100      1.396239             33.641208   
...         ...        ...       ...           ...                   ...   
2433  91.187608   4.445537  1.405213      0.858126             60.095060   
2434  90.130702   5.943936  0.472935      0.485833             51.648588   
2435  65.463178  30.550955  0.645909      0.104891             49.758980   
2436  94.625317   0.154508  0.629070      0.684251             49.880605   
2437  91.393432   3.381528  0.055305      3.149772             49.892641   

      BirthRate    Id  binnedInc_ordinal  
0      6.838710     0                0.0  
1      4.799131     1                0.0  
2      3.996826     2                3.0  
3      3.291536     3                8.0  
4      3.166561     4                0.0  
...         ...   ...                ...  
2433   5.175689  3041                9.0  
2434   4.651829  3042                3.0  
2435   4.344104  3043                2.0  
2436   6.210826  304

**[2.4]** Feature creation: This method involves creating new features from existing features that capture additional information.  We could create a new feature that calculates the percentage of the population that is not white by combining the "PctBlack", "PctAsian", and "PctOtherRace" features"

For training set

In [128]:
# Create a new feature "PctNonWhite"
df_cleaned["PctNonWhite"] = 100 - df_cleaned[["PctWhite", "PctBlack", "PctAsian", "PctOtherRace"]].sum(axis=1)


In [129]:
df_cleaned

Unnamed: 0  avgAnnCount  avgDeathsPerYear  TARGET_deathRate  \
0              0    88.000000                40             261.0   
1              1    73.000000                35             167.3   
2              2   292.000000               124             191.0   
3              3  1962.667684                 7             165.4   
4              4    43.000000                20             160.6   
...          ...          ...               ...               ...   
2433        2433   389.000000               157             185.3   
2434        2434   286.000000               117             196.7   
2435        2435   103.000000                42             204.1   
2436        2436  1962.667684                23             171.1   
2437        2437   219.000000                89             179.7   

      incidenceRate  medIncome  popEst2015  povertyPercent  studyPerCap  \
0        561.400000      29090       13352            26.8  2771.120431   
1        345.600000      29782       21903            38.8     0.000000   
2        468.400000      41955       48985            15.5     0.000000   
3        453.549422      55378        3007            11.1     0.000000   
4        349.700000      26309        8551            35.3     0.000000   
...             ...        ...         ...             ...          ...   
2433     456.600000      65485       83199             7.7     0.000000   
2434     492.400000      42477       46222            16.9   281.251352   
2435     506.700000      40339       18201            21.3     0.000000   
2436     453.549422      39764        8856            16.7     0.000000   
2437     449.100000      40801       37585            17.2     0.000000   

               binnedInc  ...  PctPublicCoverageAlone   PctWhite   PctBlack  \
0       [22640, 34218.1]  ...                    32.9  99.693045   0.044920   
1       [22640, 34218.1]  ...                    34.1  94.791383   1.649850   
2     (40362.7, 42724.4]  ...                    16.0  95.102348   1.741749   
3     (54545.6, 61494.5]  ...                    15.2  85.833870   0.933677   
4       [22640, 34218.1]  ...                    24.5  24.535525  73.223736   
...                  ...  ...                     ...        ...        ...   
2433   (61494.5, 125635]  ...                    11.3  91.187608   4.445537   
2434  (40362.7, 42724.4]  ...                    18.8  90.130702   5.943936   
2435  (37413.8, 40362.7]  ...                    20.6  65.463178  30.550955   
2436  (37413.8, 40362.7]  ...                    21.3  94.625317   0.154508   
2437  (40362.7, 42724.4]  ...                    21.8  91.393432   3.381528   

      PctAsian  PctOtherRace  PctMarriedHouseholds  BirthRate    Id  \
0     0.000000      0.000000             55.499459   6.838710     0   
1     0.063631      2.854286             52.818296   4.799131     1   
2     0.376429      0.445611             50.560800   3.996826     2   
3     0.160979      7.244044             52.565181   3.291536     3   
4     0.394100      1.396239             33.641208   3.166561     4   
...        ...           ...                   ...        ...   ...   
2433  1.405213      0.858126             60.095060   5.175689  3041   
2434  0.472935      0.485833             51.648588   4.651829  3042   
2435  0.645909      0.104891             49.758980   4.344104  3043   
2436  0.629070      0.684251             49.880605   6.210826  3045   
2437  0.055305      3.149772             49.892641   6.549178  3046   

      binnedInc_ordinal  PctNonWhite  
0                   0.0     0.262035  
1                   0.0     0.640851  
2                   3.0     2.333862  
3                   8.0     5.827431  
4                   0.0     0.450400  
...                 ...          ...  
2433                9.0     2.103516  
2434                3.0     2.966594  
2435                2.0     3.235067  
2436                2.0     3.906854  
2437                3.0     2.019963  

[2438

For testing set

In [130]:
# Create a new feature "PctNonWhite"
df_test_cleaned["PctNonWhite"] = 100 - df_test_cleaned[["PctWhite", "PctBlack", "PctAsian", "PctOtherRace"]].sum(axis=1)

**[2.5]** Scaling: This method is useful for continuous features that have different scales, such as "avgAnnCount" and "avgDeathsPerYear"and "popEst2015". It involves scaling the features so that they have similar ranges

Scalling and normalizing "avgAnnCount"

In [131]:
# For training set
from sklearn.preprocessing import MinMaxScaler

# Create a MinMaxScaler object
scaler = MinMaxScaler()

# Scale and normalize the "avgAnnCount" variable
df_cleaned['avgAnnCount'] = scaler.fit_transform(df_cleaned[['avgAnnCount']])

In [132]:
# For testing set

# Create a MinMaxScaler object
scaler = MinMaxScaler()

# Scale and normalize the "avgAnnCount" variable
df_test_cleaned['avgAnnCount'] = scaler.fit_transform(df_test_cleaned[['avgAnnCount']])

Scalling and normalizing "avgDeathsPerYear"

In [133]:
# For training set

# Create a MinMaxScaler object
scaler = MinMaxScaler()

# Scale and normalize the "avgAnnCount" variable
df_cleaned['avgDeathsPerYear'] = scaler.fit_transform(df_cleaned[['avgDeathsPerYear']])

In [134]:
# For testing set

# Create a MinMaxScaler object
scaler = MinMaxScaler()

# Scale and normalize the "avgAnnCount" variable
df_test_cleaned['avgDeathsPerYear'] = scaler.fit_transform(df_test_cleaned[['avgDeathsPerYear']])

Scalling and normalizing "popEst2015"

In [135]:
# For training set

# Create a MinMaxScaler object
scaler = MinMaxScaler()

# Scale and normalize the "avgAnnCount" variable
df_cleaned['popEst2015'] = scaler.fit_transform(df_cleaned[['popEst2015']])

In [136]:
# For testing set

# Create a MinMaxScaler object
scaler = MinMaxScaler()

# Scale and normalize the "avgAnnCount" variable
df_test_cleaned['popEst2015'] = scaler.fit_transform(df_test_cleaned[['popEst2015']])

**[2.5]** Handle outliners: Usually z-score =3 is considered as a cut-off value to set the limit. Therefore, any z-score greater than +3 or less than -3 is considered as outlier which is pretty much similar to standard deviation method

For training set

In [137]:
# calculate the z-scores for each numeric column
z_scores = np.abs((df_cleaned - df_cleaned.mean()) / df_cleaned.std())

# create a boolean mask for values that are more than 3 standard deviations from the mean
mask = (z_scores > 3).any(axis=1)

# remove the rows that contain outliers
df_cleaned = df_cleaned[~mask]

<ipython-input-137-b789c1ddc4e7>:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  z_scores = np.abs((df_cleaned - df_cleaned.mean()) / df_cleaned.std())


For testing shape

In [138]:
# calculate the z-scores for each numeric column
z_scores_test = np.abs((df_test_cleaned - df_test_cleaned.mean()) / df_test_cleaned.std())

# create a boolean mask for values that are more than 3 standard deviations from the mean
mask_test = (z_scores_test > 3).any(axis=1)

# remove the rows that contain outliers
df_test_cleaned = df_test_cleaned[~mask_test]

<ipython-input-138-2b09abcb1d66>:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  z_scores_test = np.abs((df_test_cleaned - df_test_cleaned.mean()) / df_test_cleaned.std())


**[2.6]** Removing redundant or irrelevant features: Some features may not be useful or might be correlated with other features in the dataset. Identifying and removing such features can simplify the dataset and improve model performance.

In [141]:
# For training dataframes
df_cleaned = df_cleaned.drop(columns=[ "Unnamed: 0","Geography","studyPerCap", "PctWhite", "PctBlack", "PctAsian", "PctOtherRace", "Id", "binnedInc"])

In [142]:
# For testing dataframes
df_test_cleaned = df_test_cleaned.drop(columns=[ "Unnamed: 0","Geography","studyPerCap","PctWhite", "PctBlack", "PctAsian", "PctOtherRace", "Id", "binnedInc"])

**[2.7]** Double checking the list of columns for training and testing dataframes

In [143]:
# For training dataframes
df_cleaned.columns

Index(['avgAnnCount', 'avgDeathsPerYear', 'TARGET_deathRate', 'incidenceRate',
       'medIncome', 'popEst2015', 'povertyPercent', 'MedianAge',
       'MedianAgeMale', 'MedianAgeFemale', 'AvgHouseholdSize',
       'PercentMarried', 'PctNoHS18_24', 'PctHS18_24', 'PctBachDeg18_24',
       'PctHS25_Over', 'PctBachDeg25_Over', 'PctEmployed16_Over',
       'PctUnemployed16_Over', 'PctPrivateCoverage', 'PctEmpPrivCoverage',
       'PctPublicCoverage', 'PctPublicCoverageAlone', 'PctMarriedHouseholds',
       'BirthRate', 'binnedInc_ordinal', 'PctNonWhite'],
      dtype='object')

In [144]:
# For testing dataframes
df_test_cleaned.columns

Index(['avgAnnCount', 'avgDeathsPerYear', 'TARGET_deathRate', 'incidenceRate',
       'medIncome', 'popEst2015', 'povertyPercent', 'MedianAge',
       'MedianAgeMale', 'MedianAgeFemale', 'AvgHouseholdSize',
       'PercentMarried', 'PctNoHS18_24', 'PctHS18_24', 'PctBachDeg18_24',
       'PctHS25_Over', 'PctBachDeg25_Over', 'PctEmployed16_Over',
       'PctUnemployed16_Over', 'PctPrivateCoverage', 'PctEmpPrivCoverage',
       'PctPublicCoverage', 'PctPublicCoverageAlone', 'PctMarriedHouseholds',
       'BirthRate', 'binnedInc_ordinal', 'PctNonWhite'],
      dtype='object')

**[2.8]** Extract the target variable into a variable called y and remove it from the training dataframe

In [145]:
y = df_cleaned.pop('TARGET_deathRate')

**[2.9]** Extract the target variable into a variable called y_test and remove it from the testing dataframe

In [146]:
y_test = df_test_cleaned.pop('TARGET_deathRate')

**[2.10]** Save all features from the training dataset into a variable called X


In [147]:
X = df_cleaned

**[2.11]** Save the all features from the testing dataset into a variable called X_test


In [148]:
X_test = df_test_cleaned

**[2.12]** Import train_test_split from sklearn.model_selection

In [149]:
from sklearn.model_selection import train_test_split

 Split the features and target variable into a training and validation sets

In [150]:
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42)

### 3. Assess Baseline Model

**[3.1]** Find the average of the target variable

In [151]:
y_train_mean = y_train.mean()
y_train_mean

179.80329322355922

**[3.2]** Create a numpy array called y_base filled with this value

In [152]:
y_train_base = np.full(y_train.shape, y_train_mean)
y_train_base

array([179.80329322, 179.80329322, 179.80329322, ..., 179.80329322,
       179.80329322, 179.80329322])

**[3.3]** Import the MSE and MAE from sklearn

In [153]:
from sklearn.metrics import mean_squared_error as mse
from sklearn.metrics import mean_absolute_error as mae

**[3.4]** Display the RMSE and MAE scores of this baseline model

In [154]:
print(mse(y_train, y_train_base, squared = False))
print(mae(y_train,y_train_base))

24.248625156083627
19.168893179600367


### 4. Train Initial Model with the all features

**[4.1]** Import the linear regression module from sklearn

In [155]:
from sklearn.linear_model import LinearRegression 

**[4.2]** instantiate the LinearRegression class into a variable called reg1

In [156]:
reg = LinearRegression()

**[4.3]** Task: Fit the model with the prepared data

In [157]:
reg = reg.fit(X_train,y_train)

**[4.4]** Use the trained model to predict the outcome on X_train and save them into y_preds

In [158]:
y_train_preds = reg.predict(X_train)
y_train_preds

array([179.08531299, 192.75716539, 164.99166819, ..., 177.84513664,
       179.63590508, 185.78147169])

**[4.5]** Display the RMSE and MAE scores on the training set

In [159]:
print(mse(y_train, y_train_preds, squared = False))
print(mae(y_train,y_train_preds))

17.055027219230297
12.910862079521637


**[4.6]** Display the RMSE and MAE scores on the validation set

In [160]:
reg_val = LinearRegression()
reg_val = reg_val.fit(X_val,y_val)
y_val_preds = reg_val.predict(X_val)
y_val_preds
print(mse(y_val, y_val_preds, squared = False))
print(mae(y_val,y_val_preds))

17.125852696136274
13.320562579703703


**[4.7]** Display the predictions versus the target line charts on the training set

In [161]:
perfect_line_chart_training = alt.Chart(pd.DataFrame({'y_target': y_train, 'y_predict': y_train})).mark_line(color='green').encode(
    x='y_target',
    y='y_predict'
  )
perfect_line_chart_training
predict_line_chart_training = alt.Chart(pd.DataFrame({'y_target': y_train, 'y_predict': y_train_preds})).mark_line().encode(
    x='y_target',
    y='y_predict'
  )
predict_line_chart_training
perfect_line_chart_training + predict_line_chart_training

alt.LayerChart(...)

**[4.8]** Display the predictions versus the target line charts on the validation set

In [162]:
perfect_line_chart_validation = alt.Chart(pd.DataFrame({'y_target': y_val, 'y_predict': y_val})).mark_line(color='green').encode(
    x='y_target',
    y='y_predict'
  )
perfect_line_chart_validation

predict_line_chart_validation = alt.Chart(pd.DataFrame({'y_target': y_val, 'y_predict': y_val_preds})).mark_line().encode(
    x='y_target',
    y='y_predict'
  )
predict_line_chart_validation

perfect_line_chart_validation + predict_line_chart_validation

alt.LayerChart(...)

**[4.9]** Display the RMSE and MAE scores on the testing set

In [163]:
reg_test = LinearRegression()
reg_test = reg_test.fit(X_test,y_test)
y_test_preds = reg_test.predict(X_test)
y_test_preds
print(mse(y_test, y_test_preds, squared = False))
print(mae(y_test,y_test_preds))

17.22509175371689
13.097296894063826


**[4.10]** Display the predictions versus the target line charts on the testing set

In [164]:
perfect_line_chart_testing = alt.Chart(pd.DataFrame({'y_target': y_test, 'y_predict': y_test})).mark_line(color='green').encode(
    x='y_target',
    y='y_predict'
  )
perfect_line_chart_testing

predict_line_chart_testing = alt.Chart(pd.DataFrame({'y_target': y_test, 'y_predict': y_test_preds})).mark_line().encode(
    x='y_target',
    y='y_predict'
  )
predict_line_chart_testing

perfect_line_chart_testing + predict_line_chart_testing

alt.LayerChart(...)

### 5. Train Lasso Model with the all features 


**[5.1]** Import the Lasso regression module from sklearn

In [165]:
from sklearn.linear_model import Lasso 

**[5.2]** Instantiate the LinearRegression class into a variable called lasso_reg

In [166]:
lasso_reg = Lasso()

**[5.3]** Task: Fit the model on the training set

In [167]:
lasso_reg.fit(X_train, y_train)


Lasso()

**[5.4]** Import the MSE and MAE metrics from sklearn

In [168]:
from sklearn.metrics import mean_squared_error as mse
from sklearn.metrics import mean_absolute_error as mae

**[5.5]** Display the RMSE and MAE scores on the training set

In [169]:
y_lasso_preds = lasso_reg.predict(X_train)
print(mse(y_train, y_lasso_preds, squared = False))
print(mae(y_train,y_lasso_preds))

17.434635936052086
13.251234030108126


**[5.6]** Display the RMSE and MAE scores on the validation set

In [170]:
lasso_reg_val = Lasso()
lasso_reg_val.fit(X_val, y_val)
y_lasso_preds_val = lasso_reg_val.predict(X_val)
print(mse(y_val, y_lasso_preds_val, squared = False))
print(mae(y_val,y_lasso_preds_val))

17.37291345608938
13.537728305810637


**[5.7]** Display the RMSE and MAE scores on the training set when alpha = 0.8

In [171]:
lasso_reg_train = Lasso(alpha = 0.8)
lasso_reg_train.fit(X_train, y_train)
y_lasso_preds_train = lasso_reg_train.predict(X_train)
print(mse(y_train, y_lasso_preds_train, squared = False))
print(mae(y_train,y_lasso_preds_train))

17.403522977675493
13.22196073413424


**[5.8]** Display the RMSE and MAE scores on the validation set when alpha = 0.8

In [172]:
lasso_reg_val = Lasso(alpha = 0.8)
lasso_reg_val.fit(X_val, y_val)
y_lasso_preds_val = lasso_reg_val.predict(X_val)
print(mse(y_val, y_lasso_preds_val, squared = False))
print(mae(y_val,y_lasso_preds_val))

17.332791002376524
13.499613536721323


**[5.9]** Display the RMSE and MAE scores on the training set when alpha = 0.5

In [173]:
lasso_reg_train = Lasso(alpha = 0.5)
lasso_reg_train.fit(X_train, y_train)
y_lasso_preds_train = lasso_reg_train.predict(X_train)
print(mse(y_train, y_lasso_preds_train, squared = False))
print(mae(y_train,y_lasso_preds_train))

17.360297420278457
13.176754299245028


**[5.10]** Display the RMSE and MAE scores on the validation set when alpha = 0.5


In [174]:
lasso_reg_val = Lasso(alpha = 0.5)
lasso_reg_val.fit(X_val, y_val)
y_lasso_preds_val = lasso_reg_val.predict(X_val)
print(mse(y_val, y_lasso_preds_val, squared = False))
print(mae(y_val,y_lasso_preds_val))

17.283036882883053
13.450346347925816


**[5.11]** Display the RMSE and MAE scores on the training set when alpha = 0.2

In [175]:
lasso_reg_train = Lasso(alpha = 0.2)
lasso_reg_train.fit(X_train, y_train)
y_lasso_preds_train = lasso_reg_train.predict(X_train)
print(mse(y_train, y_lasso_preds_train, squared = False))
print(mae(y_train,y_lasso_preds_train))

17.33601737909376
13.145193677177188


**[5.10]** Display the RMSE and MAE scores on the validation set when alpha = 0.2

In [176]:
lasso_reg_val = Lasso(alpha = 0.2)
lasso_reg_val.fit(X_val, y_val)
y_lasso_preds_val = lasso_reg_val.predict(X_val)
print(mse(y_val, y_lasso_preds_val, squared = False))
print(mae(y_val,y_lasso_preds_val))

17.229232538443256
13.397326924738243


###6. Train Initial Model with the all features using Ridge Model

**[6.1]** Import the Ridge regression module from sklearn

In [177]:
from sklearn.linear_model import Ridge

**[6.2]** Instantiate the Ridge model into a variable called ridge_reg and fit it on the training set

In [178]:
ridge_reg = Ridge()
ridge_reg.fit(X_train, y_train)

Ridge()

**[6.4]** Print the RMSE and MAE scores for the training set

In [179]:
y_ridge_preds = ridge_reg.predict(X_train)
print(mse(y_train, y_ridge_preds, squared = False))
print(mae(y_train,y_ridge_preds))

17.25731940987965
13.080082112489048


**[6.4]** Print the RMSE and MAE scores for the validation set

In [180]:
ridge_reg_val = Ridge()
ridge_reg_val.fit(X_val, y_val)
y_ridge_preds_val = ridge_reg_val.predict(X_val)
print(mse(y_val, y_ridge_preds_val, squared = False))
print(mae(y_val,y_ridge_preds_val))

17.178417278182632
13.367154525363581


**[6.5]** Print the RMSE and MAE scores for the training set with alpha = 0.8

In [181]:
ridge_reg_train = Ridge(alpha = 0.8)
ridge_reg_train.fit(X_train, y_train)
y_ridge_preds_train = ridge_reg_train.predict(X_train)
print(mse(y_train, y_ridge_preds_train, squared = False))
print(mae(y_train,y_ridge_preds_train))

17.249920796133036
13.074431674128363


**[6.6]** Print the RMSE and MAE scores for the validation set with alpha = 0.8

In [182]:
ridge_reg_val = Ridge(alpha = 0.8)
ridge_reg_val.fit(X_val, y_val)
y_ridge_preds_val = ridge_reg_val.predict(X_val)
print(mse(y_val, y_ridge_preds_val, squared = False))
print(mae(y_val,y_ridge_preds_val))

17.177029935086335
13.365641852283447


**[6.7]** Print the RMSE and MAE scores for the training set with alpha = 0.5

In [183]:
ridge_reg_train = Ridge(alpha = 0.5)
ridge_reg_train.fit(X_train, y_train)
y_ridge_preds_train = ridge_reg_train.predict(X_train)
print(mse(y_train, y_ridge_preds_train, squared = False))
print(mae(y_train,y_ridge_preds_train))

17.232826988568373
13.060641121293225


**[6.8]** Print the RMSE and MAE scores for the validation set with alpha = 0.5

In [184]:
ridge_reg_val = Ridge(alpha = 0.5)
ridge_reg_val.fit(X_val, y_val)
y_ridge_preds_val = ridge_reg_val.predict(X_val)
print(mse(y_val, y_ridge_preds_val, squared = False))
print(mae(y_val,y_ridge_preds_val))

17.17380741953625
13.362044367857369


**[6.9]** Print the RMSE and MAE scores for the training set with alpha = 0.2

In [185]:
ridge_reg_train = Ridge(alpha = 0.2)
ridge_reg_train.fit(X_train, y_train)
y_ridge_preds_train = ridge_reg_train.predict(X_train)
print(mse(y_train, y_ridge_preds_train, squared = False))
print(mae(y_train,y_ridge_preds_train))

17.196498299126688
13.030950263857846


**[6.8]** Display the predictions versus the target line charts on the training set

In [186]:
perfect_line_chart_training = alt.Chart(pd.DataFrame({'y_target': y_train, 'y_predict': y_train})).mark_line(color='green').encode(
    x='y_target',
    y='y_predict'
  )
perfect_line_chart_training
predict_line_chart_ridge_training = alt.Chart(pd.DataFrame({'y_target': y_train, 'y_predict': y_ridge_preds_train})).mark_line().encode(
    x='y_target',
    y='y_predict'
  )
predict_line_chart_ridge_training
perfect_line_chart_training + predict_line_chart_ridge_training

alt.LayerChart(...)

**[6.9]** Print the RMSE and MAE scores for the validation set with alpha = 0.2

In [187]:
ridge_reg_val = Ridge(alpha = 0.2)
ridge_reg_val.fit(X_val, y_val)
y_ridge_preds_val = ridge_reg_val.predict(X_val)
print(mse(y_val, y_ridge_preds_val, squared = False))
print(mae(y_val,y_ridge_preds_val))

17.165398813242245
13.351963626591244


**[6.10]** Display the predictions versus the target line charts on the validation set

In [188]:
perfect_line_chart_validation = alt.Chart(pd.DataFrame({'y_target': y_val, 'y_predict': y_val})).mark_line(color='green').encode(
    x='y_target',
    y='y_predict'
  )
perfect_line_chart_validation

predict_line_chart_ridge_validation = alt.Chart(pd.DataFrame({'y_target': y_val, 'y_predict': y_ridge_preds_val})).mark_line().encode(
    x='y_target',
    y='y_predict'
  )
predict_line_chart_ridge_validation

perfect_line_chart_validation + predict_line_chart_ridge_validation

alt.LayerChart(...)

**[6.11]** Print the RMSE and MAE scores for the testing set with alpha = 0.2

In [189]:
ridge_reg_test = Ridge(alpha = 0.2)
ridge_reg_test.fit(X_test, y_test)
y_ridge_preds_test = ridge_reg_test.predict(X_test)
print(mse(y_test, y_ridge_preds_test, squared = False))
print(mae(y_test,y_ridge_preds_test))

17.326624196322687
13.173961537494426


**[6.12]** Display the predictions versus the target line charts on the testing set

In [190]:
perfect_line_chart_testing = alt.Chart(pd.DataFrame({'y_target': y_test, 'y_predict': y_test})).mark_line(color='green').encode(
    x='y_target',
    y='y_predict'
  )
perfect_line_chart_testing

predict_line_ridge_chart_testing = alt.Chart(pd.DataFrame({'y_target': y_test, 'y_predict': y_ridge_preds_test})).mark_line().encode(
    x='y_target',
    y='y_predict'
  )
predict_line_ridge_chart_testing

perfect_line_chart_testing + predict_line_ridge_chart_testing

alt.LayerChart(...)

###7. Train Initial Model with the all features using Elastinet Model

**[7.1]** Import the ElasticNet module from sklearn

In [191]:
from sklearn.linear_model import ElasticNet 

**[7.2]** Instantiate the ElasticNet class into a variable called elasticnet_reg and fit it on the training set

In [192]:
elasticnet_reg = ElasticNet()
elasticnet_reg.fit(X_train, y_train)

ElasticNet()

**[7.3]** Print the RMSE and MAE scores for the training set

In [193]:
y_elasticnet_preds = elasticnet_reg.predict(X_train)
print(mse(y_train, y_elasticnet_preds, squared = False))
print(mae(y_train,y_elasticnet_preds))

17.385202304364952
13.194764318503559


**[7.4]** Print the RMSE and MAE scores for the validation set

In [194]:
elasticnet_reg_val = ElasticNet()
elasticnet_reg_val.fit(X_val, y_val)
y_elasticnet_preds_val = elasticnet_reg_val.predict(X_val)
print(mse(y_val, y_elasticnet_preds_val, squared = False))
print(mae(y_val,y_elasticnet_preds_val))

17.348644813294648
13.522673072126848


**[7.5]** Print the RMSE and MAE scores for the training set with alpha = 0.8

In [195]:
elasticnet_reg = ElasticNet(alpha = 0.8)
elasticnet_reg.fit(X_train, y_train)
y_elasticnet_preds = elasticnet_reg.predict(X_train)
print(mse(y_train, y_elasticnet_preds, squared = False))
print(mae(y_train,y_elasticnet_preds))

17.369713733005266
13.179874850592704


**[7.6]** Print the RMSE and MAE scores for the validation set with alpha = 0.8

In [196]:
elasticnet_reg_val = ElasticNet(alpha = 0.8)
elasticnet_reg_val.fit(X_val, y_val)
y_elasticnet_preds_val = elasticnet_reg_val.predict(X_val)
print(mse(y_val, y_elasticnet_preds_val, squared = False))
print(mae(y_val,y_elasticnet_preds_val))

17.31822195688319
13.493543294101059


**[7.7]** Print the RMSE and MAE scores for the training set with alpha = 0.5

In [197]:
elasticnet_reg = ElasticNet(alpha = 0.5)
elasticnet_reg.fit(X_train, y_train)
y_elasticnet_preds = elasticnet_reg.predict(X_train)
print(mse(y_train, y_elasticnet_preds, squared = False))
print(mae(y_train,y_elasticnet_preds))

17.34938736740848
13.158864141856448


**[7.8]** Print the RMSE and MAE scores for the validation set with alpha = 0.5

In [198]:
elasticnet_reg_val = ElasticNet(alpha = 0.5)
elasticnet_reg_val.fit(X_val, y_val)
y_elasticnet_preds_val = elasticnet_reg_val.predict(X_val)
print(mse(y_val, y_elasticnet_preds_val, squared = False))
print(mae(y_val,y_elasticnet_preds_val))

17.273782822794935
13.44608937244819


**[7.9]** Print the RMSE and MAE scores for the training set with alpha = 0.2

In [199]:
elasticnet_reg = ElasticNet(alpha = 0.2)
elasticnet_reg.fit(X_train, y_train)
y_elasticnet_preds = elasticnet_reg.predict(X_train)
print(mse(y_train, y_elasticnet_preds, squared = False))
print(mae(y_train,y_elasticnet_preds))

17.33183439369888
13.14058279673929


**[7.10]** Print the RMSE and MAE scores for the validation set with alpha = 0.2

In [200]:
elasticnet_reg_val = ElasticNet(alpha = 0.2)
elasticnet_reg_val.fit(X_val, y_val)
y_elasticnet_preds_val = elasticnet_reg_val.predict(X_val)
print(mse(y_val, y_elasticnet_preds_val, squared = False))
print(mae(y_val,y_elasticnet_preds_val))

17.22514529001715
13.401511502179453
